File: runtimes/openclaw/src/core/agt-tools/http-fetch.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

// http_fetch AGT tool registration — extracted from plugin.ts in S15.f.8.

import { routerCall } from "../router-client.js";
import { safeJson } from "../safe-json.js";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyApi = any;

export function registerHttpFetchTool(api: AnyApi): void {
  api.registerTool({
    name: "http_fetch",
    label: "HTTP Fetch (Egress Proxy)",
    description:
      "Make an HTTP request to an external URL. The request is routed through the AzureClaw security proxy which enforces blocklist (51K+ malicious domains blocked), allowlist, and learn mode. Use this for ANY external API call (Telegram, HackerNews, web APIs, etc.). Direct internet access via curl/fetch is blocked by the egress guard.",
    parameters: {
      type: "object",
      properties: {
        url: { type: "string", description: "The full URL to fetch (e.g., https://api.telegram.org/bot.../getMe)" },
        method: { type: "string", description: "HTTP method: GET, POST, PUT, DELETE. Default: GET" },
        headers: { type: "object", description: "Optional HTTP headers as key-value pairs" },
        body: { type: "string", description: "Optional request body (for POST/PUT)" },
      },
      required: ["url"],
    },
    async execute(_id: string, params: Record<string, unknown>) {
      try {
        const result = await routerCall("POST", "/egress/fetch", {
          url: params.url,
          method: (params.method as string) || "GET",
          headers: params.headers || {},
          body: params.body || undefined,
        });
        return { content: [{ type: "text", text: safeJson(result) }] };
      } catch (e: any) {
        return { content: [{ type: "text", text: `Fetch failed: ${e.message}` }] };
      }
    },
  });
}

File: runtimes/hermes/src/kars_runtime_hermes/plugin/http_fetch.py
"""http_fetch tool — Phase A1.4 (always-on).

HTTP fetch routed through the inference router's ``/egress/fetch``
endpoint. The router enforces:
  * Blocklist (always — even in Learn mode)
  * Allowlist (Strict mode)
  * Learn mode (records all new domains)
  * SSRF protection (rejects localhost/private/.local/.internal hosts)

Mirror of OpenClaw's ``runtimes/openclaw/src/core/agt-tools/http-fetch.ts``.
"""

from __future__ import annotations

import json
import logging
from typing import Any

from . import router_client

logger = logging.getLogger("kars.hermes.http_fetch")


def _http_fetch(args: dict[str, Any], **_kwargs: Any) -> str:
    url = str(args.get("url", "")).strip()
    if not url:
        return json.dumps({"error": "url is required"})
    method = str(args.get("method", "GET")).upper()
    headers = args.get("headers") or {}
    body = args.get("body") or ""

    payload = {
        "url": url,
        "method": method,
        "headers": headers if isinstance(headers, dict) else {},
        "body": str(body),
    }
    try:
        resp = router_client.call("POST", "/egress/fetch", json=payload)
    except Exception as exc:  # noqa: BLE001
        return json.dumps({"error": f"egress fetch failed: {exc}"})

    if resp.status_code >= 400:
        return json.dumps(
            {
                "error": f"egress denied (HTTP {resp.status_code})",
                "body": resp.text[:500],
            }
        )
    # Return body verbatim — caller can parse as needed.
    return resp.text[:32 * 1024]  # cap at 32KB for prompt safety


_HTTP_FETCH_SCHEMA = {
    "name": "http_fetch",
    "description": (
        "HTTP fetch routed through the kars egress proxy. Subject to "
        "blocklist (always), Learn-mode logging (default), or allowlist "
        "(Strict mode). SSRF-protected — localhost/private IPs rejected. "
        "Use for arbitrary HTTPS endpoints; for high-level web search "
        "use foundry_web_search if available."
    ),
    "parameters": {
        "type": "object",
        "properties": {
            "url": {"type": "string", "description": "Absolute URL (https:// or http://)"},
            "method": {"type": "string", "description": "GET / POST / PUT / PATCH / DELETE / HEAD (default GET)"},
            "headers": {"type": "object", "description": "Optional HTTP headers"},
            "body": {"type": "string", "description": "Optional request body (string)"},
        },
        "required": ["url"],
    },
}


def register(ctx: Any) -> None:  # noqa: ANN401
    ctx.register_tool(
        name="http_fetch",
        toolset="http_fetch",
        schema=_HTTP_FETCH_SCHEMA,
        handler=_http_fetch,
        description=_HTTP_FETCH_SCHEMA["description"],
    )
    logger.info("http_fetch registered")

